src/main/scala/com/codacy/api/client/CodacyClient.scala

```diff
@@ -21,96 +21,78 @@ class CodacyClient(apiUrl: Option[String] = None, apiToken: Option[String] = Non
   /*
    * Does an API request and parses the json output into a class
    */
-  def execute[T](request: Request[T])(implicit extractor: Extractor[T, Json]): RequestResponse[T] = {
+  def execute[T](request: Request[T])(implicit reads: Reads[T]): RequestResponse[T] = {
     get(request.endpoint) match {
-      case SuccessfulResponse(json) => SuccessfulResponse(json.as[T])
+      case SuccessfulResponse(json) => json.validate[T].fold(
+        errors => FailedResponse(JsonOps.handleDerivationFailure(errors)),
+        derived => SuccessfulResponse(derived)
+      )
       case f: FailedResponse => f
     }
   }
 
   /*
    * Does an paginated API request and parses the json output into a sequence of classes
    */
-  def executePaginated[T](request: Request[Seq[T]])(implicit extractor: Extractor[T, Json]): RequestResponse[Seq[T]] = {
+  def executePaginated[T](request: Request[Seq[T]])(implicit reads: Reads[T]): RequestResponse[Seq[T]] = {
+    implicit val paginatedResultReads: Reads[PaginatedResult[T]] = Json.reads[PaginatedResult[T]]
     get(request.endpoint) match {
       case SuccessfulResponse(json) =>
-        val nextPage = (json \ "next").as[Option[String]]
-        val nextRepos = nextPage.map {
-          nextUrl =>
-            executePaginated(Request(nextUrl, request.classType))
-        }.getOrElse(SuccessfulResponse(Seq.empty))
-
-        val values = (json \ "values").as[List[T]]
-        //.fold[RequestResponse[List[T]]](FailedResponse(s"Failed to parse json: $json"))(a => SuccessfulResponse(a))
-        RequestResponse.apply(SuccessfulResponse(values), nextRepos)
-
+        json.validate[PaginatedResult[T]].fold(
+          errors => FailedResponse(JsonOps.handleDerivationFailure(errors)),
+          {
+            case PaginatedResult(Some(nextUrl), values) =>
+              val nextRepos = executePaginated(Request(nextUrl, request.classType))
+              RequestResponse(SuccessfulResponse(values), nextRepos)
+            case PaginatedResult(None, values) =>
+              RequestResponse(SuccessfulResponse(values), SuccessfulResponse(Seq.empty))
+          }
+        )
       case f: FailedResponse => f
     }
   }
 
   /*
    * Does an API post
    */
-  def post[T](request: Request[T], value: String)(implicit extractor: Extractor[T, Json]): RequestResponse[T] = {
+  def post[T](request: Request[T], value: String)(implicit reads: Reads[T]): RequestResponse[T] = {
     val headers = tokens ++ Map("Content-Type" -> "application/json")
 
-    val body = Http.parse(s"$remoteUrl/${request.endpoint}")
-      .query(request.queryParameters)
-      .httpPost(value, headers)
-      .slurp[Char]
+    val body = Http(s"$remoteUrl/${request.endpoint}")
+      .params(request.queryParameters)
+      .headers(headers)
+      .postData(value)
+      .asString.body
```

Comment:
No check for the returned code?

Reply:
Handled by attempts of conversion of body to successful response